www.nature.com/scientificreports scientific re POrts @ Check for updates OPEN Improved genetic algorithm for | | | | ] ] multi-threshold optimization in digital pathol ' tati Tangsen Huang'%3™, Haibing Yin! & Xingru Huang'-3 This paper presents an improved genetic algorithm focused on multi-threshold optimization for image segmentation in digital pathology. By innovatively enhancing the selection mechanism and crossover operation, the limitations of traditional genetic algorithms are effectively addressed, significantly improving both segmentation accuracy and computational efficiency. Experimental results demonstrate that the improved genetic algorithm achieves the best balance between precision and recall within the threshold range of 0.02 to 0.05, and it significantly outperforms traditional methods in terms of segmentation performance. Segmentation quality is quantified using metrics such as precision, recall, and F1 score, and statistical tests confirm the superior performance of the algorithm, especially in its global search capabilities for complex optimization problems. Although the algorithm’s computation time is relatively long, its notable advantages in segmentation quality, particularly in handling high-precision segmentation tasks for complex images, are highly pronounced. The experiments also show that the algorithm exhibits strong robustness and stability, maintaining reliable performance under different initial conditions. Compared to general segmentation models, this algorithm demonstrates significant advantages in specialized tasks, such as pathology image segmentation, especially in resource-constrained environments. Therefore, this improved genetic algorithm offers an efficient and precise multi-threshold optimization solution for image segmentation, providing valuable reference for practical applications. Keywords Image segmentation, Improved genetic algorithm, Multi-threshold optimization, Global search capability, Robustness and stability This section introduces the research background and significance of digital pathology, the research objectives, and the critical role of multi-threshold optimization in pathological image segmentation. It also discusses the limitations of current threshold selection methods and the improvements proposed in this study. Research background and significance Digital pathology is a pivotal discipline in the medical field, significantly contributing to the understanding of disease onset, progression, and impact!. The precision of pathological diagnosis is integral to determining treatment plans and prognoses’. Conventionally, pathological diagnosis hinges on the subjective interpretation of pathologists, who examine tissue samples under microscopes—a process that is not only time-consuming but also susceptible to individual biases’. The advent of digital pathology technology, with its ability to digitize high- resolution images, has unlocked new avenues for computer-assisted pathological diagnosis. Image segmentation is a pivotal step in automated analysis, as it isolates areas of interest from the background, facilitating subsequent feature extraction and diagnostic analysis®. Threshold selection is at the heart of image segmentation technology, and its rationality and precision are critical to the analysis process. An appropriate threshold can precisely demarcate the boundaries of the area of interest, ensuring that subsequent analysis is based on the correct image segments. Hence, the accurate selection of thresholds in the intricate and variable landscape of pathological images is a pressing issue in digital pathology image analysis. 1School of Communication Engineering, Hangzhou Dianzi University, Hangzhou 310018, China. %School of Information Engineering, Hunan University of Science and Engineering, Yongzhou 425199, China. 3Lishui Institute of Hangzhou Dianzi University, Lishui 323000, China. ~email: 202243080072@hdu.edu.cn Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ Research motivation The intricate and varied nature of pathological images presents a significant challenge for threshold selection. Current threshold selection methodologies often draw on statistical properties of image gray-level distributions, which are effective for images with clear bimodal or multimodal distributions. However, the gray-level distribution in pathological images is more nuanced due to variations in tissue structure and staining techniques. Furthermore, pathological images typically feature multiple regions of interest, necessitating multi-threshold selection for more refined segmentation. This study seeks to devise an automated, efficient, and precise multi- threshold selection strategy that can accommodate the complexity and diversity of pathological images, thereby enhancing the accuracy and efficiency of pathological image analysis. Multi-threshold optimization using genetic algorithms for digital pathology In the realm of digital pathology, multi-threshold optimization is a key technique for achieving refined segmentation of pathological images by selecting multiple optimal thresholds. Traditional threshold selection methods, such as the Otsu method and k-means algorithm?®, provide a foundational framework but often fail to address the complexities and diversity of pathological images, especially when differentiating between various types of tissues and disease states. These methods typically rely on global image information, overlooking the importance of local features, and assume significant differences in grayscale distribution between image regions, which is often not the case in complex pathological images®. Therefore, there is a pressing need to develop more flexible and robust multi-threshold strategies that can dynamically adapt to the unique characteristics of each image, ensuring high precision in disease identification and staging. This study introduces a new multi-threshold selection strategy optimized through genetic algorithms, which simulate natural selection processes’. This strategy helps to better retain superior individuals, enhances population diversity, introduces innovative selection and crossover strategies, and facilitates the generation of superior offspring®. These iterative updates accelerate the algorithm’s convergence speed and effectively integrate advanced image features, such as texture and morphology, with traditional grayscale information. This significantly improves the accuracy and computational efficiency of segmentation. Moreover, this study has also optimized the parameters of the genetic algorithms to suit the characteristics of multi-threshold optimization. Experimental validation shows that the improved genetic algorithms perform with higher segmentation accuracy and computational efficiency in handling complex pathological images, providing an effective multi-threshold optimization method that enhances the accuracy and efficiency of pathological diagnoses. Main contributions of this paper This paper introduces an improved genetic algorithm for multi-threshold optimization in pathological image segmentation, addressing the limitations of existing methods. The main contributions are: (1) Improved Genetic Algorithm: We develop an enhanced genetic algorithm with a new selection mechanism and optimized crossover operations, significantly improving segmentation accuracy and efficiency. (2) Superior Performance: Comparative and statistical analyses show that our algorithm outperforms tradi- tional genetic algorithms and other methods in metrics like precision, recall, and F1 score. (3) Robustness and Stability: The algorithm demonstrates consistent and reliable performance across various initial conditions, making it suitable for real-world applications in medical image segmentation. These contributions provide an efficient and precise solution for pathological image segmentation, advancing computer-aided diagnosis. Overview of the paper The paper is organized as follows: Section “Related work” reviews traditional and contemporary pathological image segmentation methods, focusing on heuristic algorithms and the rationale for selecting genetic algorithms. Section “Methodology” details the dataset, image preprocessing, feature extraction, and the improved genetic algorithm for multi-threshold optimization. Section “Experimental results and analysis” presents experimental results, comparing the proposed method with traditional approaches, and provides statistical validation and performance analysis. Section “Discussion” discusses the implications, strengths, limitations, and future research directions. Section “Conclusion” concludes by summarizing the study’s contributions and the significance of the improved algorithm. Related work This section reviews traditional methods and recent advancements in pathological image segmentation, particularly the application of heuristic algorithms in image segmentation. It includes a comparative analysis of genetic algorithms versus other heuristic algorithms, highlighting the motivation for choosing genetic algorithms. Pathology image segmentation techniques Pathology image segmentation is a crucial component of computer-assisted pathological diagnosis, aimed at separating regions of interest, such as tumor areas, from the background in pathological images’. This segmentation lays the groundwork for subsequent feature extraction and quantitative analysis. With advancements in digital pathology, there has been a surge in research focused on developing automated, efficient, and accurate methods for pathology image segmentation'®. Traditional segmentation methods include threshold-based techniques like Otsu’s method and maximum entropy, region-based approaches such as region growing and the Watershed algorithm, and edge-based techniques using operators like Canny and Sobel!!. Although simple and Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ straightforward, these methods often fall short in handling the high complexity and diversity of pathological images effectively. In recent years, the evolution of deep learning, especially Convolutional Neural Networks (CNN), has significantly advanced pathology image segmentation'>. CNNs automatically learn and extract high-level semantic information from images by processing hierarchical features, substantially enhancing segmentation performance. State-of-the-art CNN-based methods like U-Net!?, SegNet!*, and DeepLab have demonstrated excellent capabilities in segmenting pathological images'>. Moreover, some studies have integrated traditional methods with deep learning, enhancing performance by combining advanced CNN-extracted features with traditional hand-crafted features to guide segmentation'®, or by using CNN outcomes as prior knowledge to refine traditional segmentation processes'’. These hybrid strategies capitalize on the strengths of both approaches, achieving superior segmentation results. Although pathological image segmentation technology has made considerable progress, there are still some challenges and difficulties, such as the heterogeneity of tumor regions, the variability of staining, and the complexity of tissue structure'®. These factors pose higher demands on segmentation algorithms, requiring further research and improvement in aspects such as algorithm robustness, adaptability, and generalization capabilities. Application of heuristic algorithms in image segmentation Heuristic algorithms are widely applied in the field of image segmentation due to their ability to handle the nonlinearity and complexity of image data'®. Traditional image segmentation methods like edge detection, region growing, and clustering often struggle with noisy, high-dimensional, and multimodal images. Heuristic algorithms, inspired by natural processes and collective behaviors, provide adaptive and flexible solutions to overcome these challenges®. Particularly in medical image segmentation, these algorithms are well-suited for optimizing threshold selection, a critical step that significantly influences the accuracy and efficiency of the segmentation process’!. Common heuristic algorithms used in image segmentation include Genetic Algorithm (GA)*, Particle Swarm Optimization (PSO)?*, Grey Wolf Optimizer (GWO)*, and Moth Flame Optimization (MFO)®. In the rapidly advancing field of digital pathology, state-of-the-art heuristic algorithms significantly improve the accuracy and efficiency of diagnostic processes in image segmentation. These algorithms continuously adapt and evolve, meeting the challenges presented by the complexity and variability of pathological images. Deep Learning-based Genetic Algorithms (DL-GAs) combine genetic algorithms with deep learning techniques®, harnessing the robust optimization capabilities of genetic algorithms and the powerful feature extraction abilities of deep neural networks. DL-GAs excel in segmenting complex pathological images and simultaneously optimize network architecture and hyperparameters to ensure a more customized and effective segmentation process”’. Enhanced Particle Swarm Optimization (EPSO) is a variant of the classic PSO?, incorporating improvements to increase convergence rates and solution quality in high-dimensional spaces, though it struggles with the randomness of particle movement, resulting in suboptimal segmentation in heterogeneous images. The Adaptive Grey Wolf Optimizer (AGWO) is designed to enhance global search capabilities and prevent premature convergence by adjusting control parameters based on the diversity of the search space and the optimization stage”, but it requires extensive parameter tuning. Quantum Moth Flame Optimization (QMFO) integrates quantum computing principles into the traditional MFO framework®’, enhancing the algorithm’s ability to manage the probabilistic nature of quantum states, though its practical implementation complexity may limit its usability in standard settings. CNNSwarm merges convolutional neural networks (CNNs) with swarm intelligence algorithms like PSO or GWO?!, allowing neural network models to evolve their weights and structures through the segmentation process to adapt to the unique characteristics of pathology images*~. However, it heavily relies on large labeled datasets, posing a challenge in medical fields where data availability is limited or privacy concerns are significant. In recent years, with the continuous development of deep learning models, the Segment Anything Model (SAM) has been proposed as a versatile segmentation model capable of handling various image segmentation tasks®>. SAM achieves efficient segmentation across different image types by leveraging a large amount of pre-training data and a powerful Convolutional Neural Network (CNN) architecture. However, SAM primarily relies on pre-training with large-scale datasets, which poses limitations in its generalization ability for specific domain image segmentation tasks, such as pathological image segmentation. Additionally, SAM’s model size is relatively large, making computational efficiency and memory consumption significant constraints in resource-limited environments. Therefore, while SAM performs exceptionally well in general image segmentation, further exploration is needed to find more suitable algorithms for multi-threshold optimization segmentation in pathological images. These algorithms collectively push the boundaries of medical image segmentation, yet each possesses inherent limitations that must be considered in practical applications to select the optimal algorithm for use in the medical image segmentation domain. Comparative analysis of Heuristic algorithms In this paper, we present an improved Genetic Algorithm (GA) and provide a comprehensive comparison with other enhanced versions of Genetic Algorithms, such as Adaptive Genetic Algorithms (AGA)**, Hybrid Genetic Algorithms (HGA)*®, and Multi-Objective Genetic Algorithms (MOGA)**. These enhanced versions incorporate features like adaptive mutation rates, hybridization with other algorithms (e.g., Particle Swarm Optimization (PSO), simulated annealing), and simultaneous optimization of multiple objectives. Additionally, we consider other heuristic algorithms like Ant Colony Optimization (ACO)?*, Differential Evolution (DE)?*, and Simulated Annealing (SA)*’, which offer unique optimization approaches and serve as valuable benchmarks for assessing the effectiveness of our improved GA. Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ Heuristic algorithms have significantly advanced the field of image segmentation, particularly in handling complex and nonlinear image data. Genetic Algorithms, Particle Swarm Optimization, Grey Wolf Optimizer (GWO), and Moth Flame Optimization (MFO) each bring distinct strengths to this domain. Genetic Algorithms are particularly noted for their robust optimization capabilities, making them well-suited to environments requiring adaptive strategies*’. However, the slow convergence rate of traditional GAs can limit their effectiveness in time-sensitive applications. To address this, our work proposes an improved GA that effectively mitigates these limitations. Particle Swarm Optimization is often favored for its simplicity and efficiency in locating global optima®*!. However, its performance can degrade in multimodal landscapes due to the inherent randomness in particle movements, which reduces its reliability in highly heterogeneous images. In contrast, the Grey Wolf Optimizer enhances global search capabilities and excels at preventing premature convergence®?, which is crucial for achieving accurate segmentation in complex images. However, GWO requires meticulous adjustment of control parameters, a process that can be tedious and time-consuming. Moth Flame Optimization, which incorporates quantum principles, offers heightened search efficiency, making it highly effective for tackling diverse and challenging segmentation tasks*’. Nevertheless, its implementation complexity may present practical challenges in standard clinical settings. Similarly, CNNSwarm, which merges convolutional neural networks (CNNs) with swarm intelligence, evolves neural network models dynamically to meet the demands of segmentation tasks**. Despite its high accuracy, CNNSwarm’s reliance on large labeled datasets can limit its practicality in data-scarce environments. While heuristic algorithms have pushed the boundaries of medical image segmentation, the choice of the most appropriate algorithm depends on specific application requirements and constraints, such as the nature of the image data, the urgency of the task, and available computational resources*. Each algorithm’s inherent strengths and weaknesses must be carefully balanced to ensure optimal performance in practical medical imaging applications. In choosing Genetic Algorithms as the core method for this study, we primarily considered their advantages in maintaining solution diversity and strong global search capabilities. Compared to PSO, GWO, and MFO, Genetic Algorithms exhibit higher adaptability and stability when tackling complex, multi-modal medical image segmentation problems'®. Additionally, the flexibility of Genetic Algorithms allows for the adjustment of crossover and mutation strategies according to the specific characteristics of the image data, leading to more precise segmentation results. Advantages of genetic algorithms in threshold selection Genetic Algorithms (GAs) have demonstrated significant advantages in multi-threshold optimization for image segmentation, with the following key benefits: (1) Capability to handle complex problems GAs eftectively handle complex problems by simulating natural se- lection and genetic mechanisms. In image segmentation, GAs encode thresholds as genes and explore the solution space through crossover and mutation operations, finding the best combination of thresholds for various image characteristics. (2) Balance between exploration and exploitation GAs provide a good balance between exploring new solutions and exploiting existing ones. They guide the search for high-quality solutions through a fitness function while maintaining population diversity, avoiding premature convergence to local optima. (3) Diversity maintenance GAs maintain the diversity of solutions throughout the evolutionary process through selection, crossover, and mutation operations. This helps explore different regions of the solution space, preventing early convergence to suboptimal solutions, especially when dealing with medical images with multi-modal characteristics. (4) Global search capability GAs have the ability to perform global searches by maintaining a diverse popu- lation, enabling a wide search range and avoiding getting stuck near local optima, which is particularly important for complex, multi-modal structures in medical images. (5) Adaptability GAs can adjust their parameters based on image content such as brightness, contrast, and texture, effectively adapting to different image features. Compared to traditional heuristic algorithms, GAs show superior performance in multi-threshold optimization and complex image segmentation tasks. The improved Genetic Algorithms (IGAs) introduced in this paper further enhance diagnostic accuracy and efficiency, and accelerate convergence speed, especially when dealing with the enormous variability and complex details of pathological images. They provide adaptive, robust, and efficient solutions, which are key reasons for their selection in this study. Methodology This section focuses on analyzing breast cancer tissue sections to quantitatively assess epithelial-mesenchymal transition (EMT) using whole slide images (WSIs). The process involves acquiring a high-quality dataset, performing rigorous image preprocessing, and extracting significant features. This section provides a comprehensive overview of these steps. Data set description The study utilizes a comprehensive dataset comprising 525 high-quality breast cancer tissue section images sourced from the Second Affiliated Hospital of Zhejiang University. These images, spanning various subtypes and stages, were meticulously prepared, stained, and scanned to ensure accuracy and relevance for the study. Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ Image preprocessing Image preprocessing is a critical step in ensuring the accuracy and reliability of subsequent analyses. In this study, two main preprocessing steps were performed on the whole slide images (WSIs) of breast cancer: background transformation and image scaling. Background transformation To enhance feature extraction and segmentation algorithms, we converted the white background of original WSI images to black, boosting image contrast and highlighting regions of interest like tumors. The segmentation result image was first binarized, simplifying it to two colors (typically black and white), using the formula: 255 if pizel value at (x,y) > threshold ” = - 1 flz,y) ={ 0 otherwise (1) Where, f(x,y) is the pixel value of the converted image at coordinates (z,y). Then, the original WSI image was multiplied pixel by pixel with the segmentation mask image. This operation can be represented by the following formula: r(z,y) x m(z,y) ) = AL Y) 2 L Y) 2 g(z,y) 555 2) Where, g(x,y) is the pixel value of the converted image at coordinates (z,y), r(z,y) is the pixel value of the original image at coordinates (z,y), and m(x,y) is the pixel value of the binarized mask image at coordinates (z,y). The image after background transformation clearly displayed the morphological and textural features of the regions of interest, providing a better foundation for subsequent feature extraction and analysis. Background transformation also helped to reduce redundant information in the image, lowering storage and transmission costs. Image scaling After completing background transformation, we scaled the WSI images. Due to their large size (typically over 10000 % 10000 pixels), direct analysis would incur significant computational overhead. We used image scaling to adjust the images to a smaller, specified size while maintaining the aspect ratio, standardizing WSI images from different sources and scanning devices for uniform feature extraction and analysis. We chose a target resolution of 2000 X 2000 pixels, balancing computational efficiency and analysis accuracy. This resolution retains detailed information without excessive computational costs. We calculated the scaling ratio by comparing the original image dimensions to the target resolution, using the minimum ratio of width or height to maintain aspect ratio, thus avoiding distortion.The formula for calculating the scaling ratio is: . min 2000 2000 3) S\ W' H where 7 is the scaling ratio, and 1/ and H are the width and height of the original image, respectively. Then, we used the bicubic interpolation algorithm to resample the original WSI images. Bicubic interpolation is a high-quality image scaling algorithm that generates new pixel values by performing cubic polynomial interpolation on the pixel values of local regions of the image, achieving smooth and continuous image scaling. The formula for bicubic interpolation is: / _ 3 3 . . . . Fay)=> >  wiwi)@+iy+]) (4) where, I(z,y) is the pixel value of the original image at point (z,y), I’ (x,y) is the pixel value of the interpolated image at point (z,y), and w (¢) and w (j) are the interpolation weights. During the implementation of image scaling, we also used efficient matrix operations and parallel computing techniques. We represented the original WSI image as a NumPy array and used optimized functions provided by image processing libraries such as OpenCV for bicubic interpolation. For large-scale WSI image datasets, we used a multi-threading or multi-processing parallel computing framework to divide the image into multiple sub- regions and process them in parallel to improve computational efficiency. After scaling, all WSI images were standardized to the same resolution, making it easier for subsequent feature extraction and analysis. Since bicubic interpolation was used, the scaled images could retain the visual quality and information content of the original images to the maximum extent, providing a reliable data foundation for EMT quantitative analysis. Feature extraction Feature extraction is essential for quantitative EMT analysis. Initially, the pixel sum and area of each preprocessed WSI image were calculated. These metrics provided fundamental morphological features. Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ Morphological feature extraction To extract fundamental features, each WSI image was first converted into a two-dimensional NumPy array, where each element represents the grayscale value of a pixel. The total pixel sum of the image was then calculated using NumPy’s sum function: . H W’ . . Pizel Sum = Z MZ j=1] (i,7) (5) where, I (7, ) represents the pixel grayscale value at position (i, j) in the image, and H and W represent the height and width of the image, respectively. To improve efficiency, we utilized the integral image technique, which stores cumulative sums of pixel values, allowing the pixel sum of any rectangular area to be computed quickly. This optimization reduces the time complexity from O(H x W)to O (1). Next, the area of each WSI image was calculated. Since all images were scaled to 2000 X 2000 pixels, the area in square pixels is: Area = H x W = 2000 x 2000 = 4,000,000 (6) To convert this to the physical area, the spatial resolution of 0.25 um/pixel was used: Physical Area = Area x (0.25 x 0.25) = 250,000 (7) The calculation of pixel sum and area extracts the basic morphological information of WSI images in an efficient and accurate manner. Although pixel sum and area are relatively simple morphological indicators, they play a significant role in EMT analysis. The pixel sum reflects the overall brightness and staining intensity of the image, which can be used to evaluate the staining quality and consistency of tissue sections. The image area, on the other hand, provides spatial scale information of the sample, which is crucial for estimating the size and invasiveness of tumors. These two indicators also lay the foundation for the extraction of texture and morphological features in subsequent analysis. Quantitative EMT analysis To quantify the EMT process, where epithelial cells transform into mesenchymal cells, we extracted the pixel ratio sequence of epithelial and stromal regions from each WSI image. A deep learning model using a U-Net architecture was employed to segment the WSI images into epithelial, stromal, and background regions. The U-Net model, known for its effectiveness in medical image segmentation, was pretrained on a large breast cancer WSI dataset and fine-tuned for our EMT dataset. This approach ensured accurate segmentation of complex morphological features. Using the trained U-Net model, each WSI image was segmented to obtain pixel-level annotations for the epithelial, stromal, and background regions. We calculated the pixel ratios: Epithelial Region Pixel Ratio (Epithelial Ratio) Calculation Formula: Ng Ry = ———+— 8 5= Ny LNy, ) Stromal Region Pixel Ratio (Mesenchymal Ratio) Calculation Formula: Ny Ry = ——— 9 M= NN 9 Where Rp represents the pixel ratio of the epithelial region, Rj; represents the pixel ratio of the stromal region, N i represents the total number of pixels in the epithelial region, and N, represents the total number of pixels in the stromal region. To capture the spatial dynamics of EMT, we used a sliding window approach. Each WSI image was divided into 256 X256 pixel blocks with a stride of 128 pixels. For each block, the epithelial and stromal pixel ratios were calculated and arranged in sequence, forming a pixel ratio sequence that reflects the local variations of EMT. The mathematical expression for the local analysis with a sliding window is: Sequence = [Rg1, Ry, Rp2, Ryo, - - -, Rpn, Ryn], where, Rg;and R)y; represent the epithelial and stromal pixel ratios of the th window, respectively. To enhance the robustness of the pixel ratio sequence, we applied a moving average smoothing technique. This reduced the impact of local noise and outliers, providing a more stable representation of the EMT process. The smoothing of the sequence can be implemented using the following moving average formula: Rsmooth (t) = ALZ k:t_ETV*l Rorigmal (k> (1)) Where, Ry00t1 (1) represents the smoothed pixel ratio at time, R,;yima (k) represents the original pixel ratio sequence, N and represents the size of the moving average window. The algorithm for extracting and smoothing the pixel ratio sequence is summarized below: Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ Input: WSI images, Breast Cancer WSI Image Dataset, EMT Dataset Output: Smoothed Pixel Ratio Sequence 1: Model « Convolutional Neural Network with U-Net architecture 2: Pretrain(Model, Breast Cancer WSI Image Dataset) 3: Transfer Learning(Model, EMT Dataset) 4: Fine-tune(Model) 5: for each WSI image do 6: Segmentation Result « Model(WSI Image) y b for each sliding window in WSI image with size 256 x256 and stride 128 do 8: Epithelial Ratio « Compute epithelial pixel ratio for the current window 9: Stromal Ratio « Compute stromal pixel ratio for the current window 10: Pixel Ratios.append((Epithelial Ratio, Stromal Ratio)) 11: End for 12: Smoothed Pixel Ratio Sequence « Moving Average(Pixel Ratios, Window Size) 13: End for 14: Return Smoothed Pixel Ratio Sequence Algorithm 1: Extracting Pixel Ratio Sequence of Epithelial and Stromal Regions This method provides a quantitative representation of EMT, capturing the dynamic and spatial heterogeneity of epithelial and stromal regions. The pixel ratio sequence is a valuable feature for EMT analysis, aiding in classification, grading, and prognosis. Optimal threshold selection Problem definition and mathematical modeling In data analysis and pattern recognition, selecting an optimal threshold to categorize data is crucial. This paper proposes an optimal threshold selection method using a genetic algorithm to maximize the segmentation effect of two sequences. Given two real number sequences S; and Sy, which contain n and m elements, respectively: Sl = {1‘1, X, ?> xn} and SQ - {yla Y2, ?7 ym} Define the threshold ¢ as a real number. For a given threshold ¢, we define two proportion functions: Ry (t): The proportion of elements in the sequence S; that are greater than the threshold ¢. Rs (t): The proportion of elements in the sequence S5 that are less than the threshold ¢. The fitness function F'(t) is the sum of these proportions: F(t)=Ri(t)+ Ry (t) (11) The optimization problem is to find the threshold ¢ that maximizes F (¢): t = argmng (t) = argmaz (Ri(t)+ Ro(t)) (12) The goal of this optimization problem is to find a threshold ¢*such that the sum of the proportion of elements in the sequence S; that are greater than the threshold and the proportion of elements in the sequence Ss that are less than the threshold is maximized. This threshold can be considered as the threshold that achieves the best segmentation effect on the two sequences. Since the threshold ¢ takes a continuous range of values, this is a continuous optimization problem. In the following sections, we will introduce how to effectively solve this problem using an improved genetic algorithm and elaborate on the steps and implementation of the algorithm. Multi-threshold optimization based on Improved Genetic Algorithm Genetic algorithms (GA) are suitable for solving complex optimization problems through natural selection mechanisms. Here’s how the improved GA works for multi-threshold optimization: (1) Encoding scheme Each threshold is represented by a real number, forming an array of real numbers (chro- mosome). (2) Initial population generation Generate a random set of initial threshold combinations. (3) Population evaluation Use a fitness function based on inter-class variance to evaluate each threshold com- bination: 2 2 2 ot =pilpy = p) +palps —p) (13) where, 1 ; and p 4 are the means of the two classes, p; and p, are their probabilities. (4) Selection Operation: Select individuals with higher fitness for the next generation using a “monarchical strategy” selection mechanism. (5) Crossover Operation: Randomly select two individuals and perform crossover to generate new threshold combinations, enhancing population diversity. Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ (6) Mutation Operation: Introduce new genetic variations by randomly altering some bits of individuals, pre- venting premature convergence. (7) Improvement Strategy Use elite preservation and other strategies to improve the algorithm’s convergence speed and global search capability. (8) Termination Condition Set conditions such as the maximum number of iterations or convergence of the fitness function value. Once met, the algorithm ends, and the optimal thresholds are outputted., otherwise, return to step 2. (9) Decoding Decode the optimal chromosome into actual thresholds for image segmentation. This improved genetic algorithm effectively handles multi-threshold selection, By introducing improved strategies and flexibly designing the fitness function, the genetic algorithm can better adapt to different problem scenarios, enhancing segmentation accuracy by avoiding local optima and ensuring a comprehensive search of the solution space. Experimental results and analysis This section explores the effectiveness of the improved genetic algorithm for multi-threshold optimization in pathological image segmentation. It provides an overview of the data augmentation, model training, evaluation metrics, and a detailed analysis of algorithm performance, including comparison with traditional methods and statistical validation. Data augmentation and model training Data augmentation To enhance the model’s adaptability to new data, we used extensive data augmentation techniques via the ImageDataGenerator class, including: (1) Random rotation Rotating images within a random range to simulate different viewpoints and improve generalization. (2) Horizontal flipping Flipping images along the vertical axis to increase diversity without changing content. (3) Scale scaling Changing image sizes to simulate different resolutions, enhancing generalization. These techniques enriched the training data, helping the model learn robust feature representations. Model training We adopted a transfer learning strategy, initializing our model with pretrained weights. Key steps included: (1) Weight initialization Using pretrained model weights to initialize the new model. (2) Freezing layers Freezing the initial layers and fine-tuning the top layers. (3) Training Using the gradient descent algorithm and the .fit_generator() method for training, which includes online data augmentation. This strategy accelerates convergence and enhances the model’s recognition and generalization capabilities. Model evaluation We used several metrics to evaluate the model’s performance: (1) Sensitivity Measures the ability to identify positive cases. TP Sensitivity = ———— 14 YTTPFFN (14) (2) Specificity Measures the ability to identify negative cases. TN Specificity = ————— 15 pectficity = FN T Fp (15) (3) Accuracy Overall classification accuracy. TP+TN Accuracy = —————————+ 16 YCTPYTN+ FP+ FN (16) (4) Intersection over Union (IOU) Evaluates segmentation accuracy. TP IOU = ——————— 17 TP+ FP+ FN (17) (5) Dice coefficient Another segmentation performance metric. 2x TP Dice Coef ficient = ————————— 18 1 2x TP+ FP+FN (18) Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ g 055 Jow o Jom o Table 1. Performance metrics summary. The performance of the model on different datasets BN Training set W Verification set 0.8 N Test set 5 0.6 £ g 0.4 g 0.2 0.0 & & A ) \Q\ r & & ¢ & o Index Fig. 1. Comparison of model performance on different datasets. I [ S (- [ S [ S L S (R [ N TR R [ S [ S . S [ A [ R N U S [ [ B [* S [ S o Ll s fw [w s Table 2. Comparison of Segmentation effects under different thresholds between the Improved Genetic Algorithm and the original genetic algorithm. The Dice coeflicient is another metric for assessing segmentation performance, similar to IOU but giving higher weight to the intersection. By quantitatively calculating these metrics, we can obtain scores that reflect the model’s performance in different aspects. The calculation of performance metrics is shown in Table 1. These metrics indicate the model performs well across all datasets, with high sensitivity and specificity, and good IOU and Dice coefficients, important for segmentation tasks. We also performed qualitative analysis by visualizing prediction results to identify potential misclassifications and assess sensitivity to different sample types. The visualization results are shown in Fig. 1. Through comprehensive quantitative and qualitative evaluations, we obtained valuable insights for model improvement, confirming its strong performance and generalization capabilities. Threshold selection results of the improved genetic algorithm In multi-threshold optimization algorithms, the selection of thresholds directly impacts the outcome of image segmentation. This section details the comparison of segmentation effects under different thresholds for the improved genetic algorithm presented in this paper and the original genetic algorithm. Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ Comparison of Precision for Two Threshold Selection Methods Improved GA ‘ g o - 0.70 0.75 0.80 0.85 0.90 0.95 1.00 Precision Fig. 2. Performance comparison between the improved GA and the original GA. Segmention gty Gy Table 3. Performance comparison of different threshold selection algorithms. Comparison of Segmentation effects under different thresholds Quantitative analysis To evaluate the segmentation effects of the improved genetic algorithm compared to the original genetic algorithm, we used metrics such as precision, recall, and F1 score. Table 2 presents these metrics for various thresholds (0.01 to 0.10). These results validate the effectiveness of the improvements made to the genetic algorithm, demonstrating that the modified approach is superior in image segmentation tasks when compared to the original algorithm. Statistical tests To determine the significance of performance differences, we used the t-test and Wilcox- on rank-sum test. t-test is a parametric test comparing the means of two samples. Wilcoxon rank-sum test (Mann-Whitney U test) is a nonparametric test comparing the medians of two samples. The statistical tests confirm that the performance difference is significant, with a p-value of 0.00195, indicating that the improved genetic algorithm performs significantly better than the original. Figure 2 presents a visual comparison of the performance of Improved GA (the improved genetic algorithm) and Original GA (the original genetic algorithm). In Fig. 2, the visualization depicts the precision distribution of both methods across various threshold values. By conducting statistical tests, we can ascertain that the performance disparity between the different threshold selection methodologies is statistically significant. The computed p-value is 0.00195, which is substantially lower than the conventional significance level of 0.05. This low p-value suggests that the observed performance difference is highly unlikely to have occurred by chance alone, indicating a significant difference in performance between the two methods. Therefore, the statistical test results affirm that the performance difference is highly significant, underscoring the exceptionally good performance of the improved method. Performance comparison with other threshold selection methods We compared the improved genetic algorithm with other methods: Histogram Thresholding, SVM, and PSO. The comparison metrics include segmentation quality, computation time, memory consumption, robustness, and flexibility. The comparison of the above evaluation indicators is shown in Table 3. The improved genetic algorithm demonstrates superior segmentation quality and robustness, although it requires more computation time and memory. Its balance between accuracy and efficiency makes it suitable for complex image segmentation tasks. In conclusion, the Improved GA’s performance in image segmentation, as showcased through the comparison with other threshold selection methods, demonstrates its practical efficacy and competitive advantages for use in real-world image processing tasks. Additionally, we have also created box plots and bar graphs to further demonstrate the outstanding advantages of the improved GA algorithm presented in this paper. The box plot is shown in Fig. 3, and the bar chart is shown in Fig. 4. Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ Comparison of Computing Time Comparison of Segmentation Quality 40 1.00 30 0.95 . 2 g g 0.90 o 0 —_— o © 0.80 0 1 2 3 [} 1 2 3 Time (seconds) Quality Metric (e.g., Accuracy) Fig. 3. Box plots comparing computation time and segmentation quality. Average Computing Time per Method Average Segmentation Quality per Method 30 25 08 3 = 20 5 g g 06 s 1 E ; 04 10 g 0.2 5 0 0.0 Improved GA  Histogram SVM PSO Improved GA  Histogram SVM PSO Method Method Fig. 4. Bar graphs displaying average computation time and segmentation quality. Table 4. Performance comparison between SAM and the improved genetic algorithm. By examining both Figs. 3 and 4, it becomes clear that the Improved GA has a comprehensive advantage. While it may not have the shortest computation time, its exceptional performance in segmentation quality makes it suitable for complex image segmentation tasks and applications that require high precision. In medical imaging, where complex images often require high-precision segmentation, the Improved GA demonstrates a significant advantage over the other three typical algorithms (Table 4). Comparison with segment anything model (SAM) To further validate the effectiveness of the improved genetic algorithm in pathology image segmentation, we introduced the Segment Anything Model (SAM) as a comparison model. SAM, a general-purpose segmentation model based on large-scale pre-training, was tested on the same dataset as the genetic algorithm, with key metrics such as segmentation quality, computation time, and memory consumption compared. The experimental results indicate that while SAM performs well, particularly in flexible applications, it lags behind the improved genetic algorithm in segmentation quality, especially when dealing with complex pathology images. SAM’s longer computation time and higher memory consumption also highlight the advantages of the improved genetic algorithm in scenarios where high precision and resource efficiency are critical. Additionally, SAM'’s generalization ability may be somewhat limited in specific applications, whereas the improved genetic algorithm is better suited to the demands of specialized tasks. Comparison with other improved GAs and Heuristic algorithms To thoroughly evaluate the performance of our improved GA, we compared it against several other improved versions of GAs, including Adaptive Genetic Algorithms (AGA), Hybrid Genetic Algorithms (HGA), and Multi- Objective Genetic Algorithms (MOGA). Additionally, we compared it with other popular heuristic algorithms such as Ant Colony Optimization (ACO), Differential Evolution (DE), and Simulated Annealing (SA). Table 5 Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ Segmentton gty (e E e L (i R - S T T (1 O (S = S S S . B Table 5. Performance comparison of different genetic algorithms and heuristic algorithms. Performance changes and convergence speed 0.55 0.50 % 045 T E ¢ 0.40 = g £ 035 = ' H] =8 0.30 0.25 —— The optimal fitness value of each generation population 0.20 -~~~ Global optimal performance indicators z - 6 8 10 Iterations Fig. 5. Performance changes and convergence speed during iteration. presents the performance metrics for these algorithms, including segmentation quality (measured by accuracy and F1 score), computation time, and memory consumption. As shown in Table 5, the Improved GA outperforms other algorithms in terms of segmentation quality and computational efficiency. While Adaptive GA and Hybrid GA offer competitive accuracy and F1 scores, our improved GA demonstrates faster computation times and lower memory consumption. The performance of other heuristic algorithms like ACO, DE, and SA also underscores the strengths of our approach, particularly in achieving high precision in complex segmentation tasks. The inclusion of comparisons with other improved versions of Genetic Algorithms (GAs) and heuristic algorithms highlights the distinct advantages of our proposed approach. While algorithms like Adaptive GA, Hybrid GA, and Multi-Objective GA show promising results, particularly in specific applications, our improved GA offers a better balance between segmentation quality and computational efficiency, which is critical in pathology image analysis. The comparison with other heuristic algorithms, such as Ant Colony Optimization (ACO), Differential Evolution (DE), and Simulated Annealing (SA), further reinforces the effectiveness of our method. Despite the robustness of these algorithms in general optimization tasks, they fall short in the specific demands of multi- threshold optimization in digital pathology, where precision and computational efficiency are paramount. Performance analysis of the algorithm Performance analysis is crucial to evaluate the improved genetic algorithm in multi-threshold optimization tasks. The analysis focuses on convergence speed, global optimality, and stability. Performance changes and convergence speed To assess the algorithm’s performance and convergence, we record the optimal fitness value in each generation. Improvement in fitness value over iterations indicates the algorithm’s progression towards the optimal solution. A stable increase or plateau in the fitness value shows effective convergence, while significant fluctuations suggest potential issues like premature convergence or local optima traps. Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ Comparison of final performance for different runs 0.6 0.5 £ S S B 0.4 [ 8 0.3 E £ a = 0.2 = [V 0.1 0.0 Runl Run2 Run3 Run4 Run5 Runs Fig. 6. Comparison of final performance across. The convergence speed, reflected by the number of iterations needed to reach a predetermined performance standard, is crucial for practical applications. Fast convergence is highly desirable. By plotting the relationship between performance and iterations, we visualize the convergence speed of the algorithm. The performance changes and convergence speed during iteration are shown in Fig. 5. Different runs In Fig. 5, the solid line represents the optimal fitness value per generation, and the dashed line shows the global optimal performance metric. The improved genetic algorithm shows a steady upward trend in performance, indicating good convergence speed without getting trapped in local optima. Global optimality To verify the algorithm’s ability to find the global optimal solution, we run the algorithm multiple times and compare the optimal solutions. Consistent or similar solutions across different runs suggest global optimality, indicating the algorithm is not trapped in local optima. The global optimality after multiple runs is depicted in Fig. 6. Figure 6 shows consistently high final performance metrics across all runs, approaching a certain optimal value with minimal differences. This indicates the algorithm’s capability to find high-quality solutions under various initial conditions, demonstrating robustness and stability. Stability Stability is assessed by analyzing the algorithm’s performance under different initialization conditions and random seeds. A stable algorithm shows similar performance across different runs, with minimal variation indicating good stability. Significant performance fluctuations suggest sensitivity to initialization or randomness. The experimental results for the stability of the algorithm in this paper are presented in Fig. 7. Figure 7 demonstrates similar convergence trends and small fluctuations in performance curves across different initialization conditions, showing the algorithm’s consistent performance. The final performance metrics are close to the global optimal value, confirming the algorithm’s stability and reliability in practical applications. Time complexity analysis To comprehensively understand the performance of the Improved Genetic Algorithm (Improved GA), a detailed analysis of its time complexity has been conducted. Time complexity is a critical metric for evaluating how the execution time of an algorithm scales with the size of the input. In the context of genetic algorithms, the primary time-consuming operations are selection, crossover, mutation, and fitness evaluation. (1) Selection Operation: The “Monarch Strategy” selection mechanism has a time complexity of O(N), where N is the size of the population. (2) Crossover Operation: The optimized crossover operation also exhibits a time complexity of O(N), as it involves processing each individual in the population. (3) Mutation Operation: The mutation operation also has a time complexity of O(N). (4) Fitness Evaluation: Fitness evaluation is the most time-intensive part of the algorithm, with its complexity depending on the complexity of the image segmentation task and the implementation of the fitness func- Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ Performance fluctuations of different initialization and random seeds o Initializationl Sy | — Initialization2 —— Initialization3 —— Initialization4 0409 ___ initializations x (V] © = 035 [ H] v = © £ .g 0.30 2] Q. 0.25 0.20 1.0 1.5 2.0 2.5 3.0 35 4.0 4.5 5.0 Iterations Fig. 7. Comparison of performance fluctuations with different initializations and random seeds. Time Complexity Analysis of Improved Genetic Algorithm 1e10 22 41943040000 S 20 40 % 35 e 209715 30 o 104857 25 g e S . 20 § E 15 F S 14 ot 0000 05 1.2 200 26214 180 i & 1.0 0000 e ¥ 60 120 ¥ 80 [S) 100 100 & " Popuy, 120 140 80 &° ’ ton 5,5, 160 6o & &N 180 200 0.6 Fig. 8. The 3D plot of the time complexity analysis of the improved genetic algorithm. tion. The complexity of fitness evaluation is denoted as F(N), where N represents the number of pixels or features in the image. Considering these factors, the overall time complexity of the Improved GA can be expressed as O (G * N * F(N)), where G is the number of generations, N is the population size, and F(N) is the complexity of the fitness evaluation function. To visually represent the time complexity, an experimental analysis was conducted, and a 3D plot was generated to illustrate the time complexity under various combinations of population size and number of generations. Each point in the plot corresponds to a specific combination of population size and generation, with the color intensity reflecting the magnitude of the time complexity. Text labels were added to provide direct numerical values for each point. The 3D plot of the Time Complexity Analysis of the Improved Genetic Algorithm is shown in Fig. 8. Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ Fulladgorithm [0520 _ [09  Jost 0 | e A L L L Nommon [0 s Jow @ Table 6. Comparison of ablation experiment performance. Performance Comparison of Algorithm Variants 52 %0 - 87 .. 86 88 o 87 0 89 gg 80 60 ;fw 5 [ & 0 20 W Accuracy Em Recall IR F1 Score ° Full Algorithm No Selection No Crossover No Mutation Fig. 9. Performance comparison of algorithm variants. The experimental results indicate that the time complexity increases with both the population size and the number of generations. This is intuitive, as each individual in the population requires processing through selection, crossover, mutation, and fitness evaluation, and these operations are directly proportional to the population size and the number of generations. In practical scenarios, the choice of population size and number of generations involves a trade-off between computational efficiency and solution quality. For applications where high-quality solutions are paramount, the Improved GA’s superior performance justifies its higher computational demands. The time complexity analysis provides a profound understanding of the Improved GA’s performance, aiding in informed decision-making when designing and applying genetic algorithms in practice. Ablation experiment In order to further evaluate the algorithm in this article, ablation studies were conducted, and the ablation experiments in this article aim to analyze and evaluate the influence of different components in the algorithm. The primary experiment compared the full algorithm with three variants: one without a selection strategy, one without a crossover operation, and one without a mutation operation. These variants were created by removing specific operations from the complete algorithm to explore their impact on performance. After conducting the ablation experiment, performance data of each variant were obtained, and the results of the ablation experiment were presented in a table (as shown in Table 6) and a graph (as shown in Fig. 9) to visually compare the performance differences of different variants. The results clearly show that the selection strategy significantly impacts the algorithm’s performance. Removing the selection strategy led to notable declines in accuracy, recall, and F1 Score, highlighting its critical role in maintaining algorithmic efficiency and effectiveness. Additionally, the removal of crossover and mutation operations also impacted performance, albeit to a lesser extent. The slight decrease in performance following the removal of the crossover operation suggests its role in introducing new solutions into the algorithm, aiding in exploration and diversity. The relatively minor impact of removing the mutation operation implies that while it helps explore new solutions, its core contribution is not as significant as the selection strategy and crossover operation in the current parameter setting and problem definition. Through ablation experiment, the study not only validates the effectiveness of the algorithm but also provides deep insights into how each component collaborates, offering valuable perspectives for further research and development. Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ Total Running Time vs. Number of Generations for Genetic Algorithm 1000 800 % 600 % 400 200 ‘ 0 0 20 40 60 80 100 Number of Generations Fig. 10. Total running time vs. number of generations. Running time analysis In the evaluation of the genetic algorithm presented, a critical aspect was the assessment of its computational efficiency, focusing on the relationship between the number of generations and the total running time required for execution. This analysis is pivotal for understanding the scalability of the algorithm and for practical implementations where computational resources may be limited. The running times were measured using a standard system, conducted on a computing environment with a CPU of Intel Core i7-8700 K and a GPU of NVIDIA GeForce RTX 2080Ti. The genetic algorithm was configured with a population size of 100 individuals, with the complexity of the fitness function moderated to reflect typical use cases in computational biology and optimization problems. The running time over a range of generations from 1 to 100 was examined to provide a comprehensive view of the algorithm’s performance under different load conditions. The time taken for each generation, which included the evaluation of the fitness function for each individual in the population, was measured. The average time per generation was consistently around 10 s. This duration remained consistent across all tests, indicating stable computational demands per generation, irrespective of the sequence in the overall run. The total running time exhibited a linear relationship with the number of generations, as illustrated in the Fig. 10. For a single generation, the algorithm required 10 s, scaling up to 1000 s for 100 generations. This linearity is an expected outcome given that each generation involves a fixed amount of computation — specifically, the evaluation of the fitness function for each individual. The linear increase in running time with the number of generations confirms the algorithm’s predictable behavior under varying operational scales. This predictability is advantageous for users needing to estimate the time requirements for larger problems or longer runs. However, it also highlights the importance of optimizing the efficiency of the fitness function evaluation and possibly reducing the population size or the number of generations for very large-scale problems to manage the computational cost effectively. The conducted running time analysis underscores the genetic algorithm’s consistency and scalability. Convergence process analysis The convergence process of the genetic algorithm under study is crucial for understanding its effectiveness in solving optimization problems and reaching satisfactory solutions within reasonable time frames. The genetic algorithm was tested on a series of benchmark optimization problems commonly used in the field. The focus was on both the rate of convergence and the quality of the solutions obtained. Parameters were adjusted to evaluate their impact on the convergence process. Convergence was assessed by monitoring the fitness scores of the population over successive generations. The primary metric for convergence was the change in average fitness of the population, alongside the best fitness score observed. The experiments were repeated multiple times to ensure statistical significance of the results. (1) Rate of Convergence: The algorithm demonstrated a rapid improvement in fitness scores within the initial generations, followed by a gradual approach towards a plateau. This suggests that the majority of significant improvements were made early in the process, with subsequent generations refining the solutions. Rate of Convergence of the Genetic Algorithm as shown in Fig. 11. (2) Impact of Parameters: Variations in mutation rates significantly influenced the convergence rate. Higher mutation rates generally led to faster convergence but at the risk of instability in the solution quality. Conversely, lower mutation rates stabilized the convergence process but required more generations to reach optimal fitness levels. Impact of Mutation Rates on Convergence of the Genetic Algorithm as shown in Fig. 12. The convergence analysis shows that the genetic algorithm’s efficiency can be improved by tailoring its operations and selection methods. Convergence is closely tied to the right parameter settings, which vary by problem. Understanding this process helps predict the needed generations for a good solution and adjust Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ Rate of Convergence of the Genetic Algorithm 10 -——W- § ) I ' ----I ) I ----I L - ] o 10 20 30 40 50 Generation Fig. 11. Rate of convergence of the genetic algorithm. Impact of Mutation Rates on Convergence of the Genetic Algorithm - Mutation Rate =000 | | . 1 = Mutation Rate = 0.1 ’ A 0.7 1 = Mutation Rate = 0.3 ¥ /N 0.6 o N Bl A | il T Nl - Nl /‘- N JZ&z1 1 1 1 1 - -z___—- o 10 20 30 40 50 Generation Fig. 12. Impact of mutation rates on convergence. e N e LSRN e Sl i - . ey J e s ki . il s R < 53 33 B 7 ) ] Sy e % et '.~—‘1i‘?~ ; \ "ﬂ :, it ‘ & o TR ) A e, - FETE e o 2o W) . :j‘ 37, ST O oy %, h 'u\ G s g o b T A j £ A | A ¢ Shenal ey Y R *\~ ’ “.: " —, | ", L 3 '-"f,' .,' .& ':: “:": ._‘, O ] ,;1“-:._(3 L N N : :e A Z.:-z\{k- TR 3‘.\“.§§. ; 19 ( N‘ Y ‘v*, N . 'wt. p // J’ =3 \v ‘A : e B g ; 5 ._ ’ ” 5 LAy - oy VL 2 R¥ ] [ ] ¢ ";“'?@\‘ ‘..{?;“\-\ 29 “ i*g"'::--» :. ¥ s RN = X ‘% e 3 Fig. 13. Representative samples of original pathology images from the dataset. Scientific Reports|  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ ] gl o". .'.s%_y 6. 2 ?%é% ‘:‘ﬁ ; < R e L e e oy el AR Eoe BV, Y gty Gl W :; S o ,%; {"’“}%. i e .‘...:&:'. ‘J‘- ' § v ;’u?/ W‘,&?ﬁd. K : o - " X ::'4-. 3 @%E £ AR S - - r S § W WERIE™ TN y 2 ok 2 4 e At . o b Fig. 14. Segmentation results using the improved genetic algorithm. parameters for better performance. This analysis underscores the importance of convergence in genetic algorithms and paves the way for future improvements. Samples of data set images and segmentation results To provide a clear visualization of the dataset and the effectiveness of the segmentation results, we have included representative samples of the images used in our experiments and the corresponding segmentation outcomes produced by the improved genetic algorithm. Figure 13 shows examples of original pathology images from the dataset. These images represent various original pathology images from the breast cancer dataset, illustrating various tissue structures and. staining patterns. Figure 14 presents the segmentation results corresponding to the images in Fig. 13. The images display the results of applying the improved genetic algorithm, with different tissue regions accurately segmented and color- coded to distinguish between epithelial, stromal, and background areas. Discussion This study introduces an enhanced genetic algorithm tailored for optimizing multi-threshold selection in image segmentation. By refining selection and crossover strategies from the original genetic algorithm, significant improvements in segmentation accuracy and efficiency are achieved. Experimental findings demonstrate the superiority of the enhanced genetic algorithm over the original, particularly evident in thresholds ranging from 0.02 to 0.05, where it achieves peak F1 scores, effectively balancing precision and recall. Evaluation metrics such as precision, recall, and F1 score are utilized to quantify segmentation quality. Results indicate that as the threshold increases, both algorithms exhibit higher precision but lower recall, suggesting a tendency to predict only highly confident positive classes, potentially missing some true positives. Statistical tests including t-tests and Wilcoxon tests confirm the significant performance advantages of the enhanced genetic algorithm compared to other threshold selection methods. Its robust global search capability enables effective handling of complex optimization challenges, particularly beneficial in tasks requiring high-precision segmentation of intricate images. While acknowledging longer computation times as a limitation, the study underscores the algorithm’s robustness and stability across diverse initial conditions, affirming its reliability in practical applications. Future research could focus on optimizing computation efficiency, expanding application domains beyond image segmentation, and broadening datasets to enhance generalizability. The comparative analysis with the SAM model further validates the superiority of the improved genetic algorithm in pathology image segmentation. While SAM excels in general image segmentation, its performance in handling domain-specific tasks, such as pathology image segmentation, shows certain limitations, particularly in computational efficiency and memory consumption. The experimental results reveal that the improved genetic algorithm not only outperforms SAM in segmentation quality but also offers significant advantages in resource-constrained environments. This suggests that, although general-purpose models like SAM possess strong adaptability, specialized algorithms can provide more efficient and precise solutions for highly specialized tasks. In summary, this study introduces an eflicient and accurate multi-threshold optimization solution for image segmentation through an enhanced genetic algorithm. Its superior performance metrics and robustness underscore its potential as a valuable tool in various real-world applications, particularly in fields demanding precise image segmentation. Conclusion This study presents an advanced genetic algorithm specifically tailored for multi-threshold optimization in digital pathology image segmentation. By enhancing traditional genetic algorithms with a “monarch strategy” Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ selection mechanism and optimized crossover operations, our approach successfully addresses common challenges such as slow convergence and local optima entrapment. The experimental results demonstrate that the improved genetic algorithm consistently delivers superior performance in segmentation quality, precision, and computational efficiency when compared to both traditional genetic algorithms and a range of other heuristic methods, including Adaptive, Hybrid, and Multi-Objective Genetic Algorithms, as well as Ant Colony Optimization, Differential Evolution, and Simulated Annealing. Additionally, our algorithm outperforms the Segment Anything Model (SAM) in the specific context of digital pathology, highlighting its robustness and adaptability to complex image analysis tasks. While the algorithm demands slightly more computational resources, its benefits in achieving high-precision segmentation make it a valuable tool for advancing digital pathology and computer-aided diagnosis. Future work will focus on further optimizing the algorithm’s efficiency and expanding its applicability to a broader range of medical imaging challenges. Data availability The datasets used and analysed during the current study available from the corresponding author on reasonable request. Received: 2 May 2024; Accepted: 16 September 2024 Published online: 28 September 2024 References 1. Liron, P. & Bui, M. Digital pathology: An overview. Surg. Pathol. Clin. 14(3), 407-421 (2021). 2. Metin, N., Gurcan & Tomaszewski, J. E. Introduction to digital pathology. Annu. Rev. Biomed. Eng. 22(1), 313-326 (2020). 3. Ravikanth Papineni, S., Ghosh & Viswanath, K. Applications of digital pathology: An overview. Biomolecules 12 (3), 416 (2022). 4. Khang, A. & Sivaraman, K. A. Big data, cloud computing and IoT: Tools and applications/edited. J. Future Revol. Comput. Sci. Commun. Eng. 4(4), 599-602 (2023). 5. Gharehchopogh, F. et al. Slime mould algorithm: A comprehensive survey of its variants and applications. Arch. Comput. Methods Eng. 30(4), 2683-2723 (2023). 6. Hao, S. et al. Salp swarm algorithm with iterative mapping and local escaping for multi-level threshold image segmentation: A skin cancer dermoscopic case study. J. Comput. Des. Eng. 10 (2), 655-693 (2023). 7. Tao Lei, X. et al. Nandi. Adaptive morphological reconstruction for seeded image segmentation. IEEE Trans. Image Process. 28(11), 5510-5523 (2019). 8. Wang, Z. et al. Multi-threshold segmentation of breast cancer images based on improved dandelion optimization algorithm. J. Supercomputing 80(3), 3849-3874 (2024). 9. Ren, L. et al. Multi-level thresholding segmentation for pathological images: Optimal performance design of a new modified differential evolution. Comput. Biol. Med. 148, 105910 (2022). 10. Zhang, J. et al. Intelligent speech technologies for transcription, disease diagnosis, and medical equipment interactive control in smart hospitals: A review. Comput. Biol. Med. 153, 106517 (2023). 11. Xing, J. et al. Boosting whale optimizer with quasi-oppositional learning and gaussian barebone for feature selection and COVID-19 image segmentation. J. Bionic Eng. 20(2), 797-818 (2023). 12. Kumar, N. et al. A dataset and a technique for generalized nuclear segmentation for computational pathology. IEEE Trans. Med. Imaging 36(7), 1550-1560 (2017). 13. Gaber, A. et al. Automatic classification of fatty liver disease based on supervised learning and genetic algorithm. Appl. Sci. 12(1), 521 (2022). 14. Chen, L. C., Papandreou, G., Kokkinos, 1., Murphy, K. & Yuille, A. L. DeepLab: Semantic image segmentation with deep convolutional nets, atrous convolution, and fully connected CRFs. IEEE Trans. Pattern Anal. Mach. Intell. 40 (4), 834-848 (2018). 15. Dhal, K. et al. Randomly attracted rough firefly algorithm for histogram based fuzzy image clustering. Knowl. Based Syst. 216, 106814 (2021). 16. Korsuk Sirinukunwattana, S. E. et al. Rajpoot. Locality sensitive deep learning for detection and classification of nuclei in routine colon cancer histology images. IEEE Trans. Med. Imaging 35 (5), 1196-1206 (2016). 17. Wei, J. et al. Genetic U-Net: Automatically designed deep networks for retinal vessel segmentation using a genetic algorithm. IEEE Trans. Med. Imaging 41(2), 292-307 (2021). 18. Taino, D. F. et al. Analysis of cancer in histological images: Employing an approach based on genetic algorithm. Pattern Anal. Appl. 24, 483-496 (2021). 19. Chen, J. et al. Multi-threshold image segmentation based on an improved differential evolution: Case study of thyroid papillary carcinoma. Biomed. Signal Process. Control 85, 104893 (2023). 20. Zhao, S. et al. Performance optimization of salp swarm algorithm for multi-threshold image segmentation: Comprehensive study of breast cancer microscopy. Comput. Biol. Med. 139, 105015 (2021). 21. Guo, H. et al. Multi-threshold image segmentation based on an improved salp swarm algorithm: Case study of breast cancer pathology images. Comput. Biol. Med. 168, 107769 (2024). 22. Houssein, E. H. et al. Accurate multilevel thresholding image segmentation via oppositional snake optimization algorithm: Real cases with liver disease. Comput. Biol. Med. 169, 107922 (2024). 23. Yang, Z. & Wu, A. A non-revisiting quantum-behaved particle swarm optimization based multilevel thresholding for image segmentation. Neural Comput. Appl. 32(16), 12011-12031 (2020). 24. Sahoo, S. K. et al. Self-adaptive moth flame optimizer combined with crossover operator and fibonacci search strategy for COVID-19 CT image segmentation. Expert Syst. Appl. 227, 120367 (2023). 25. Yu, X. & Wu, X. Ensemble grey wolf optimizer and its application for image segmentation. Expert Syst. Appl. 209, 118267 (2022). 26. Hagq, E. et al. An MRI-based deep learning approach for efficient classification of brain tumors. J. Ambient Intelli. Human. Comput. 1-22 (2023). 27. Junyue, C. et al. Breast cancer diagnosis using hybrid AlexNet-ELM and chimp optimization algorithm evolved by nelder-mead simplex approach. Biomed. Signal Process. Control 85, 105053 (2023). 28. Houssein, E. H., Marwa, M., Emam, Abdelmgeid, A. & Ali Improved manta ray foraging optimization for multi-level thresholding using COVID-19 CT images. Neural Comput. Appl. 33(24), 16899-16919 (2021). 29. Chen, J. et al. Renal pathology images segmentation based on improved cuckoo search with diffusion mechanism and adaptive beta-hill climbing. J. Bionic Eng. 20(5), 2240-2275 (2023). 30. Liu, L. et al. Performance optimization of differential evolution with slime mould algorithm for multilevel breast cancer image segmentation. Comput. Biol. Med. 138, 104910 (2021). Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ 31. Chen, Y. et al. Multi-threshold image segmentation using a multi-strategy shuffled frog leaping algorithm. Expert Syst. Appl. 194, 116511 (2022). 32. Yang, X( et ;l. Multi-threshold image segmentation for melanoma based on Kapur’s entropy using enhanced ant colony optimization. Front. Neuroinform. 16, 1041799 (2022). 33. Kirillov, A. et al. Segment anything. In Proceedings of the IEEE/CVF International Conference on Computer Vision 4015-4026 2023). 34. g\/[azulowski, M. A. et al. Segment anything model for medical image analysis: An experimental study. Med. Immage. Anal. 89, 102918 (2023). 35. Han, S. é(Li, ))( An improved adaptive genetic algorithm. In SHS Web of Conferences Vol. 140, 01044 (EDP Sciences, 2022). 36. Oh, L. S, Lee, J. S. & Moon, B. R. Hybrid genetic algorithms for feature selection. IEEE Trans. Pattern Anal. Mach. Intell. 26(11), 1424-1437 (2004). 37. Konak, A., Coit, D. W. & Smith, A. E. Multi-objective optimization using genetic algorithms: A tutorial. Reliab. Eng. Syst. Saf. 91(9), 992-1007 (2006). 38. Murata, T. & Ishibuchi, H. MOGA: Multi-objective genetic algorithms. In IEEE International Conference on Evolutionary Computation Vol. 1 289-294 (IEEE Piscataway, 1995). 39. Blum, C. Ant colony optimization: Introduction and recent trends. Phys. Life Rev. 2(4), 353-373 (2005). 40. Yang, X. et al. Multi-level threshold segmentation framework for breast cancer images using enhanced differential evolution. Biomed. Signal Process. Control 80, 104373 (2023). 41. Ranjbarzadeh, R. et al. Breast tumor localization and segmentation using machine learning techniques: Overview of datasets, findings, and methods. Comput. Biol. Med. 152, 106443 (2023). 42. Wang, X. et al. Medical image segmentation using PCNN based on multi-feature grey wolf optimizer bionic algorithm. J. Bionic Eng. 18, 711-720 (2021). 43. Vishnoi, S., Jain, A. K. & Sharma, P. K. An efficient nuclei segmentation method based on roulette wheel whale optimization and fuzzy clustering. Evol. Intel. 14(3), 1367-1378 (2021). 44. Rai, R., Das, A. & Dhal, K. G. Nature-inspired optimization algorithms and their significance in multi-thresholding image segmentation: An inclusive review. Evol. Syst. 13(6), 889-945 (2022). 45. Chen, K. et al. Lung computed tomography image segmentation based on U-Net network fused with dilated convolution. Comput. Methods Programs Biomed. 207, 106170 (2021). Author contributions T.H. formulated the problem definition, designed the mathematical modeling, and wrote the main manuscript text. H.Y. developed the improved genetic algorithm, conducted the statistical analysis, and participated in draft- ing the manuscript. X.H. designed and conducted the experiments, prepared all figures, and participated in the analysis of experimental results. All authors reviewed the manuscript. Funding This work was supported by the Key R&D projects of the Ministry of Science and Technology of China (No. 2023YFB4502800), in part by the Natural Science Foundation of China (No. 62102147), in part by the National Natural Science Foundation of China (No. 41771406), in part by the Key Scientific Research Foundation of Hunan Provincial Department of Education (No. 23A0575), in part by the Hunan Provincial Natural Science Foundation (Nos. 2024]]7184, 2024]]7187, 2022]]30275) . Declarations Competing interests The authors declare no competing interests. Ethical approval This study adheres to the guidelines set forth in the "Regulations onEthical Review of Human Biomedical Research" (Order No. 11 of theNational Health Commission of the People’s Republic of China) and the "DataSecurity Law of the People’s Republic of China". The breast cancerpathological images utilized in this research have undergone thoroughde-identification. In accordance with relevant regulations, my study does notrequire additional ethical approval and informed consent. The study protocol and the use of de-identified data were reviewed and approved by the Institutional Review Board (IRB) of Hangzhou Danzi University. The IRB determined that the study was exempt from requiring full IRB review because it used existing, de- identified data. Informed consent The study utilized de-identified breast cancer whole-slide images (WSIs) sourced from public datasets. As these data were de-identified and publicly accessible, the requirement for informed consent was waived by the Institutional Review Board (IRB) of Hangzhou Danzi University. This waiver was granted on the basis that the data were already de-identified and did not contain any personal health information. The authors have ensured that all patient data used in this study were handled with confidentiality and privacy. No personal health information was accessed or used in the preparation of this manuscript. Additional information Correspondence and requests for materials should be addressed to T.H. Reprints and permissions information is available at www.nature.com/reprints. Publisher’s note Springer Nature remains neutral with regard to jurisdictional claims in published maps and institutional affiliations. Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

www.nature.com/scientificreports/ Open Access This article is licensed under a Creative Commons Attribution-NonCommercial-NoDerivatives 4.0 International License, which permits any non-commercial use, sharing, distribution and reproduction in any medium or format, as long as you give appropriate credit to the original author(s) and the source, provide a link to the Creative Commons licence, and indicate if you modified the licensed material. You do not have permission under this licence to share adapted material derived from this article or parts of it. The images or other third party material in this article are included in the article’s Creative Commons licence, unless indicated otherwise in a credit line to the material. If material is not included in the article’s Creative Commons licence and your intended use is not permitted by statutory regulation or exceeds the permitted use, you will need to obtain permission directly from the copyright holder. To view a copy of this licence, visit http://creativecommons.org/ licenses/by-nc-nd/4.0/. © The Author(s) 2024 Scientific Reports |  (2024) 14:22454 | https://doi.org/10.1038/s41598-024-73335-6 nature portfolio

